aloe/tools.py

```diff
@@ -93,3 +95,54 @@ def guess_types(data):  # pylint:disable=too-complex
 
     #  give up
     return data
+
+
+def hook_not_reentrant(func):
+    """
+    Decorate a hook as unable to be reentered while it is already in the
+    stack.
+
+    Any further attempts to enter the hook before exiting will be replaced
+    by a no-op.
+
+    This is generally useful for step hooks where a step might call
+    :meth:`Step.behave_as` and trigger a second level of step hooks i.e.
+    when displaying information about the running test.
+    """
+
+    # pylint:disable=protected-access
+    func._entered = False
```

Comment:
If I put it inside inner it will be a different reference, if I put it outside inner it would be a one reference for all invocations. In which case I could use a dict, but I'm not sure why this approach is any worse.